### PR TITLE
Only microbatch in train mode + del logits to save memory

### PR DESCRIPTION
**Pull request by sdtblck**

### CODE REVIEW

**Comment by sdtblck** [comment]
pytest failure is fixed in the latest commit to main branch

**Review by leogao2 — APPROVED** [review]
